presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -671,15 +672,16 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
 
         private WindowNode.Specification canonicalizeAndDistinct(WindowNode.Specification specification)
         {
-            ImmutableMap.Builder<Symbol, SortOrder> orderings = ImmutableMap.builder();
+            LinkedHashMap<Symbol, SortOrder> orderings = new LinkedHashMap<>();
```

Comment:
IMO storing this as simple collections is too simple. We might loose this semantic you are trying to impose in a future (because of some unrelated change).
What about extracting some entity (like Assignments), where you could say in javadoc what is a contract etc.

Like below you are using `putIfAbsent`, it is very easy to use `put` instead and your contract will be not satisfied.

Reply:
The right solution for that would be to introduce structure (class) representing sorting with appropriate checks and accessor methods.
Although, this patch is simple to address one problem without refactoring sizable part of code.

I tend to disagree that this implementation can be easily broken because we have tests protecting it. So when you change it, you will notice that you broke a test.

Also, we were relying on `LinkedHashMap` semantics before, e.g. in https://github.com/prestodb/presto/pull/8822/files#diff-21d31cb9d487772413eff4f67b8555deR696 where we few lines later use `keySet` relying on the fact it'll be ordered set: https://github.com/prestodb/presto/pull/8822/files#diff-21d31cb9d487772413eff4f67b8555deR747.

Reply:
Ok. lets leave this refactoring as a followup.